File: Examples/ConektaCardToken-Swift/Conekta/Token.m
//
//  Token.m
//  Conekta
//
//  Created by Julian Ceballos on 10/16/15.
//
//

#import <Foundation/Foundation.h>
#include "Token.h"

@implementation Token

- (id) init
{
    [self setResourceURI:@"/tokens"];
    return self;
}

- (id) initWithCard: (Card*) card
{
    self = [super init];
    if (!self) return nil;
    
    [self setCard: card];
    
    return self;
}

- (void) createWithSuccess: (void(^)(NSDictionary *))successHandler andError: (void(^)(NSError*))errorHandler
{
    NSString *url = [NSString stringWithFormat:@"%@%@", [self baseURI], [self resourceURI]];
    NSURL *resource_url = [NSURL URLWithString: url];
    NSMutableURLRequest *urlRequest = [NSMutableURLRequest requestWithURL:resource_url];
    
    [urlRequest setHTTPMethod:@"POST"];
    
    [urlRequest addValue:[NSString stringWithFormat:@"Basic %@", [self apiKeyBase64:[self publicKey]]] forHTTPHeaderField:@"Authorization"];
    [urlRequest addValue:@"application/json" forHTTPHeaderField: @"Content-type"];
    [urlRequest addValue:@"application/vnd.conekta-v0.3.0+json" forHTTPHeaderField: @"Accept"];
    [urlRequest addValue:@"{\"agent\":\"Conekta Conekta iOS SDK\"}" forHTTPHeaderField: @"Conekta-Client-User-Agent"];

    [urlRequest setHTTPBody:[[self card] asJSONData]];
    
    [self deviceCollect];
    
    Connection *connection = [[Connection alloc] initWithRequest:urlRequest];
    [connection requestWithSuccess: successHandler andError: errorHandler];
}

- (NSString *) apiKeyBase64: (NSString *)api_key
{
    NSData *plainData = [api_key dataUsingEncoding:NSUTF8StringEncoding];
    NSData *apiKeyBase64Data = [plainData base64EncodedDataWithOptions: NSDataBase64Encoding64CharacterLineLength];
    NSString *returnValue = [[NSString alloc] initWithData:apiKeyBase64Data encoding:NSUTF8StringEncoding];
    return returnValue;
}

- (void) deviceCollect
{
    DeviceCollectorSDK *dc = [[DeviceCollectorSDK alloc] initWithDebugOn:NO];
    [dc setDelegate:self];
    [dc setMerchantId:@"205000"];
    [dc setCollectorUrl:@"https://api.conekta.io/fraud_providers/kount/logo.htm"];
    [dc collect:[self deviceFingerprint]];
}

- (void) onCollectorStart
{
}

- (void) onCollectorSuccess
{
}

- (void) onCollectorError: (int32_t)errorCode withError: (NSError *)error
{
    NSLog(@"%@", error);
}


@end

File: Examples/ConektaCardToken-Swift/Conekta/Token.swift
//
//  Token.swift
//  ConektaCardToken-Swift
//
//  Created by Ricardo Michel Reyes Martínez on 7/21/16.
//  Copyright © 2016 Conekta. All rights reserved.
//

import Foundation

class Token
{
    var baseURI = ""
    var publicKey = ""
    var resourceURI = ""
    
    var card: Card?
    var deviceFingerprint = ""
    
    init()
    {
        self.resourceURI = "/tokens"
    }
    
    convenience init(card: Card)
    {
        self.init()
        self.card = card
    }
    
    func create(completionHandler: (response: AnyObject) -> (), errorHandler: (error: NSError) -> ())
    {
        if let url = NSURL(string: "\(self.baseURI)\(self.resourceURI)")
        {
            let request = NSMutableURLRequest(URL: url)
            request.HTTPMethod = "POST"
            
            if let apiKey = self.apiKeyBase64(self.publicKey)
            {
                request.setValue("Basic \(apiKey)", forHTTPHeaderField: "Authorization")
            }
            
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.setValue("application/vnd.conekta-v0.3.0+json", forHTTPHeaderField: "Accept")
            request.setValue("{\"agent\":\"Conekta iOS SDK\"}", forHTTPHeaderField: "Conekta-Client-User-Agent")
            request.HTTPBody = self.card?.jsonSerialization()
            
            let connection = Connection(request: request)
            connection.request(completionHandler, errorHandler: errorHandler)
        }
    }
    
    func apiKeyBase64(apiKey: String) -> String?
    {
        let plainData = apiKey.dataUsingEncoding(NSUTF8StringEncoding)
        
        if let apiKeyBase64Data = plainData?.base64EncodedDataWithOptions(.Encoding64CharacterLineLength)
        {
            return NSString(data: apiKeyBase64Data, encoding: NSUTF8StringEncoding) as? String
        }
        
        return nil
    }
}